Persist invited users

andresilva opened the pull request:
Fix #3.

An option `--invited-list` is added which takes a file path where github ids of invited users will be persisted. This is a plain text file where one github id (a number) is stored per line (this makes it easy to manually edit).

andresilva committed:
add struct to track invited github ids

andresilva committed:
track previously invited users

andresilva committed:
fix tests

andresilva committed:
add docs to --invited-list

andresilva commented:
I didn't actually test this since the setup to do so isn't easy. I think it should work but would appreciate if someone who has access to the existing infrastructure could actually test it.

andresilva committed:
rename Invites to Invited

andresilva committed:
create invited file if it doesn't exist

andresilva committed:
parse invited list option with from_os_str

cole-h commented on src/op_sync_team.rs:
```diff
@@ -202,19 +211,24 @@ pub fn sync_team(
                     "github-name" => format!("{}", github_name),
                 ));
-                if pending_invites.contains(&github_name) {
+                if pending_invites.contains(&github_name) || invited.contains(&github_id) {
```
I think it might be valuable to distinguish between the two.

cole-h commented on src/op_sync_team.rs:
```diff
@@ -268,13 +282,15 @@ pub fn sync_team(
                 removals.inc();
                 info!(logger, "Removing user from the team");
                 if do_it_live {
+                    invited.remove(&github_id);
```
We should only remove them from the `invited` list if they are successfully removed (down below, around line 320 with the current changes, could be changed into a `match` on the result of `rt.block_on(...)`).

andresilva replied:
Doh, of course :)

cole-h commented on src/op_sync_team.rs:
```diff
@@ -202,19 +211,24 @@ pub fn sync_team(
+                        // keep track of the invitation locally so that we don't
+                        // spam users that have already been invited and rejected
+                        // the invitation
+                        invited.add(github_id.clone());
```
Same as above (users should only be added if they were successfully invited). There's already a `match` statement ready for you to add on to, for this one.

cole-h reviewed (changes requested):
Overall LGTM, just a few comments.

Thanks for taking initiative on this! I had been meaning to get around to it, but... y'know, life, work, holidays, yadda-yadda.

andresilva committed:
only add/remove invited if successful

andresilva committed:
distinguish between pending invite / rejected invite

andresilva commented:
I wanted to be added to the NixOS organization and was told that wouldn't happen until this issue was closed 🙈. Thanks for the quick review @cole-h, I think I addressed all comments.

grahamc commented on src/op_sync_team.rs:
```diff
@@ -202,14 +211,18 @@ pub fn sync_team(
             }
         }
         match action {
-            TeamAction::Add(github_name, handle) => {
+            TeamAction::Add(github_name, github_id, handle) => {
                 let logger = logger.new(o!(
                     "nixpkgs-handle" => format!("{}", handle),
                     "github-name" => format!("{}", github_name),
                 ));
+
                 if pending_invites.contains(&github_name) {
                     noops.inc();
                     debug!(logger, "User already has a pending invitation");
+                } else if invited.contains(&github_id) {
+                    noops.inc();
+                    debug!(logger, "User was already invited previously (since there's no pending invitation we can assume the user rejected the invite)");
```
Previously we tracked "noops" as a proxy for how many people were pending invitations. Can you add two metrics for this,

below:

```
    let removals =
        register_int_counter!("rfc39_team_sync_removals", "Total team removals").unwrap();
```

```
    let pending_invitations =
        register_int_counter!("rfc39_team_sync_invite_pending", "Total pending team invitations pending").unwrap();
    let previously_invited =
        register_int_counter!("rfc39_team_sync_previously_invited", "Total users not invited because we know we invited them already").unwrap();
```

and then:

```suggestion
                if pending_invites.contains(&github_name) {
                    noops.inc();
                    pending_invitations.inc();
                    debug!(logger, "User already has a pending invitation");
                } else if invited.contains(&github_id) {
                    noops.inc();
                    previously_invited.inc();
                    debug!(logger, "User was already invited previously (since there's no pending invitation we can assume the user rejected the invite)");
```

and finally extend this bit of code:

```
        let logger = logger.new(o!(
            "dry-run" => dry_run,
            "github-id" => format!("{}", github_id),
            "changed" => additions.get() + removals.get(),
            "additions" => additions.get(),
            "removals" => removals.get(),
            "noops" => noops.get(),
            "errors" => errors.get(),
        ));
```

to include these two metrics?

grahamc reviewed (commented):
Looks pretty good to me, I'll give this a test shortly. Thank you for working on this!

One thing I'm thinking is maybe the Invited struct should receive a logger and print log messages in the error cases which right now just `?`: IO errors especially are hard to track down, and parsing can be a bit fraught, though this parsing is super simple. What do you think?

I also think we should probably export a gauge of how many entries are loaded and saved, for monitoring purposes.

Does this sound good? I can help with these things if you'd like!

grahamc commented on src/cli.rs:
```diff
@@ -70,6 +70,12 @@ pub struct SyncTeamParams {
+    /// guarantees that users that have been previously invited and rejected
+    /// will not keep getting spammed.
+    #[structopt(long = "invited-list", parse(from_os_str))]
+    pub invited_list: Option<PathBuf>,
```
Let's make this mandatory: we can't run this program safely without it.

andresilva committed:
make invited list file mandatory

andresilva committed:
log errors on invited list file operations

andresilva committed:
add metrics for pending invitations and previously invited

andresilva committed:
add gauges for invited list loaded / saved

andresilva commented:
@grahamc Thanks for the review. Just pushed some more changes which I think should address all your suggestions.

grahamc commented:
It looks like it isn't quite right. Check this out, after this run the invitations list is empty and these two metrics are notable:

```
# HELP rfc39_invited_list_loaded Number of github ids loaded from the previously invited list
# TYPE rfc39_invited_list_loaded gauge
rfc39_invited_list_loaded 0
# HELP rfc39_invited_list_saved Number of github ids saved to the previously invited list
# TYPE rfc39_invited_list_saved gauge
rfc39_invited_list_saved 0
```

```
[nix-shell:~/nixpkgs]$ ../rfc39/result/bin/rfc39 --dump-metrics --metrics-delay=1 --metrics-addr=0.0.0.0:9190 --credentials /run/keys/rfc39-credentials.nix --maintainers ./maintainers/maintainer-list.nix sync-team NixOS 3345117 --limit 1 --invited-list ./invitations
Dec 28 16:29:56.164 INFO Listening on 0.0.0.0:9190, thread: metrics, module: rfc39:223
Dec 28 16:29:56.164 INFO Loading maintainer information, absolute: /root/nixpkgs/maintainers/maintainer-list.nix, from: ./maintainers/maintainer-list.nix, module: rfc39:73
Dec 28 16:29:56.201 INFO Loading GitHub authentication information from "/run/keys/rfc39-credentials.nix", module: rfc39:82
Dec 28 16:29:56.552 INFO Syncing team, team_id: 3345117, team_name: Nixpkgs Maintainers, exec-mode: SyncTeam, module: rfc39::op_sync_team:136
Dec 28 16:29:56.552 INFO Fetching current team members, team_id: 3345117, team_name: Nixpkgs Maintainers, exec-mode: SyncTeam, module: rfc39::op_sync_team:141
Dec 28 16:30:10.839 INFO Adding user to the team, github-name: thelegy, nixpkgs-handle: thelegy, errors: 0, noops: 12, previously-invited: 0, pending-invitations: 0, removals: 0, additions: 0, changed: 0, github-id: 3105057, dry-run: false, exec-mode: SyncTeam, module: rfc39::op_sync_team:251
Dec 28 16:30:11.285 INFO Hit maximum change limit, errors: 0, noops: 12, previously-invited: 0, pending-invitations: 0, removals: 0, additions: 1, changed: 1, github-id: 311580, dry-run: false, exec-mode: SyncTeam, module: rfc39::op_sync_team:230
metrics:
 # HELP rfc39_begin_seconds Execution started time
# TYPE rfc39_begin_seconds gauge
rfc39_begin_seconds 1609172996
# HELP rfc39_github_add_user Time to add a GitHub user to a team
# TYPE rfc39_github_add_user histogram
rfc39_github_add_user_bucket{le="0.005"} 0
rfc39_github_add_user_bucket{le="0.01"} 0
rfc39_github_add_user_bucket{le="0.025"} 0
rfc39_github_add_user_bucket{le="0.05"} 0
rfc39_github_add_user_bucket{le="0.1"} 0
rfc39_github_add_user_bucket{le="0.25"} 0
rfc39_github_add_user_bucket{le="0.5"} 1
rfc39_github_add_user_bucket{le="1"} 1
rfc39_github_add_user_bucket{le="2.5"} 1
rfc39_github_add_user_bucket{le="5"} 1
rfc39_github_add_user_bucket{le="10"} 1
rfc39_github_add_user_bucket{le="+Inf"} 1
rfc39_github_add_user_sum 0.294310402
rfc39_github_add_user_count 1
# HELP rfc39_github_add_user_failures Number of failed attempts to add a user
# TYPE rfc39_github_add_user_failures counter
rfc39_github_add_user_failures 0
# HELP rfc39_github_call_count Code-level calls to GitHub API methods (not a count of actual calls made to GitHub.)
# TYPE rfc39_github_call_count counter
rfc39_github_call_count 5
# HELP rfc39_github_get_invitations Time to fetch invitations
# TYPE rfc39_github_get_invitations histogram
rfc39_github_get_invitations_bucket{le="0.005"} 0
rfc39_github_get_invitations_bucket{le="0.01"} 0
rfc39_github_get_invitations_bucket{le="0.025"} 0
rfc39_github_get_invitations_bucket{le="0.05"} 0
rfc39_github_get_invitations_bucket{le="0.1"} 0
rfc39_github_get_invitations_bucket{le="0.25"} 1
rfc39_github_get_invitations_bucket{le="0.5"} 1
rfc39_github_get_invitations_bucket{le="1"} 1
rfc39_github_get_invitations_bucket{le="2.5"} 1
rfc39_github_get_invitations_bucket{le="5"} 1
rfc39_github_get_invitations_bucket{le="10"} 1
rfc39_github_get_invitations_bucket{le="+Inf"} 1
rfc39_github_get_invitations_sum 0.144528513
rfc39_github_get_invitations_count 1
# HELP rfc39_github_get_team Time to fetch a team
# TYPE rfc39_github_get_team histogram
rfc39_github_get_team_bucket{le="0.005"} 0
rfc39_github_get_team_bucket{le="0.01"} 0
rfc39_github_get_team_bucket{le="0.025"} 0
rfc39_github_get_team_bucket{le="0.05"} 0
rfc39_github_get_team_bucket{le="0.1"} 0
rfc39_github_get_team_bucket{le="0.25"} 0
rfc39_github_get_team_bucket{le="0.5"} 1
rfc39_github_get_team_bucket{le="1"} 1
rfc39_github_get_team_bucket{le="2.5"} 1
rfc39_github_get_team_bucket{le="5"} 1
rfc39_github_get_team_bucket{le="10"} 1
rfc39_github_get_team_bucket{le="+Inf"} 1
rfc39_github_get_team_sum 0.330118487
rfc39_github_get_team_count 1
# HELP rfc39_github_get_team_failures Number of failed attempts to get a team
# TYPE rfc39_github_get_team_failures counter
rfc39_github_get_team_failures 0
# HELP rfc39_github_get_team_invitation_failures Number of failed attempts to get a team's pending invitations
# TYPE rfc39_github_get_team_invitation_failures counter
rfc39_github_get_team_invitation_failures 0
# HELP rfc39_github_get_team_members Time to fetch team members
# TYPE rfc39_github_get_team_members histogram
rfc39_github_get_team_members_bucket{le="0.005"} 0
rfc39_github_get_team_members_bucket{le="0.01"} 0
rfc39_github_get_team_members_bucket{le="0.025"} 0
rfc39_github_get_team_members_bucket{le="0.05"} 0
rfc39_github_get_team_members_bucket{le="0.1"} 0
rfc39_github_get_team_members_bucket{le="0.25"} 0
rfc39_github_get_team_members_bucket{le="0.5"} 0
rfc39_github_get_team_members_bucket{le="1"} 0
rfc39_github_get_team_members_bucket{le="2.5"} 0
rfc39_github_get_team_members_bucket{le="5"} 0
rfc39_github_get_team_members_bucket{le="10"} 0
rfc39_github_get_team_members_bucket{le="+Inf"} 1
rfc39_github_get_team_members_sum 14.140950884
rfc39_github_get_team_members_count 1
# HELP rfc39_github_get_team_members_failures Number of failed attempts to get a team's members
# TYPE rfc39_github_get_team_members_failures counter
rfc39_github_get_team_members_failures 0
# HELP rfc39_github_get_user Time to fetch a GitHub user
# TYPE rfc39_github_get_user histogram
rfc39_github_get_user_bucket{le="0.005"} 0
rfc39_github_get_user_bucket{le="0.01"} 0
rfc39_github_get_user_bucket{le="0.025"} 0
rfc39_github_get_user_bucket{le="0.05"} 0
rfc39_github_get_user_bucket{le="0.1"} 0
rfc39_github_get_user_bucket{le="0.25"} 1
rfc39_github_get_user_bucket{le="0.5"} 1
rfc39_github_get_user_bucket{le="1"} 1
rfc39_github_get_user_bucket{le="2.5"} 1
rfc39_github_get_user_bucket{le="5"} 1
rfc39_github_get_user_bucket{le="10"} 1
rfc39_github_get_user_bucket{le="+Inf"} 1
rfc39_github_get_user_sum 0.15160996
rfc39_github_get_user_count 1
# HELP rfc39_github_get_user_failures Number of failed attempts to get a user
# TYPE rfc39_github_get_user_failures counter
rfc39_github_get_user_failures 0
# HELP rfc39_github_invitation_count Currently invited users
# TYPE rfc39_github_invitation_count gauge
rfc39_github_invitation_count 1
# HELP rfc39_github_remove_user Time to remove a GitHub user from a team
# TYPE rfc39_github_remove_user histogram
rfc39_github_remove_user_bucket{le="0.005"} 0
rfc39_github_remove_user_bucket{le="0.01"} 0
rfc39_github_remove_user_bucket{le="0.025"} 0
rfc39_github_remove_user_bucket{le="0.05"} 0
rfc39_github_remove_user_bucket{le="0.1"} 0
rfc39_github_remove_user_bucket{le="0.25"} 0
rfc39_github_remove_user_bucket{le="0.5"} 0
rfc39_github_remove_user_bucket{le="1"} 0
rfc39_github_remove_user_bucket{le="2.5"} 0
rfc39_github_remove_user_bucket{le="5"} 0
rfc39_github_remove_user_bucket{le="10"} 0
rfc39_github_remove_user_bucket{le="+Inf"} 0
rfc39_github_remove_user_sum 0
rfc39_github_remove_user_count 0
# HELP rfc39_github_remove_user_failures Number of failed attempts to remove a user
# TYPE rfc39_github_remove_user_failures counter
rfc39_github_remove_user_failures 0
# HELP rfc39_github_team_member_count Fetched team members
# TYPE rfc39_github_team_member_count gauge
rfc39_github_team_member_count 996
# HELP rfc39_github_username_id_mismatch Number of maintainers not added because of out of date usernames, due to a mismatched ID
# TYPE rfc39_github_username_id_mismatch gauge
rfc39_github_username_id_mismatch 0
# HELP rfc39_invited_list_loaded Number of github ids loaded from the previously invited list
# TYPE rfc39_invited_list_loaded gauge
rfc39_invited_list_loaded 0
# HELP rfc39_invited_list_saved Number of github ids saved to the previously invited list
# TYPE rfc39_invited_list_saved gauge
rfc39_invited_list_saved 0
# HELP rfc39_maintainer_missing_key_github Maintainers missing a github handle.
# TYPE rfc39_maintainer_missing_key_github gauge
rfc39_maintainer_missing_key_github 42
# HELP rfc39_maintainer_missing_key_github_id Maintainers missing a github_id.
# TYPE rfc39_maintainer_missing_key_github_id gauge
rfc39_maintainer_missing_key_github_id 42
# HELP rfc39_maintainer_nix_load_failure Failures to load maintainers.nix
# TYPE rfc39_maintainer_nix_load_failure counter
rfc39_maintainer_nix_load_failure 0
# HELP rfc39_op_failure_counter Execution failed
# TYPE rfc39_op_failure_counter counter
rfc39_op_failure_counter 0
# HELP rfc39_op_panic_counter Execution of the operation panicked
# TYPE rfc39_op_panic_counter counter
rfc39_op_panic_counter 0
# HELP rfc39_op_suceess_counter Execution completed without fault.
# TYPE rfc39_op_suceess_counter counter
rfc39_op_suceess_counter 1
# HELP rfc39_stop_seconds Execution stopped time
# TYPE rfc39_stop_seconds gauge
rfc39_stop_seconds 1609173011
# HELP rfc39_team_sync_additions Total team additions
# TYPE rfc39_team_sync_additions counter
rfc39_team_sync_additions 1
# HELP rfc39_team_sync_change_limit Total number of additions and changed allowed in a single run
# TYPE rfc39_team_sync_change_limit gauge
rfc39_team_sync_change_limit 1
# HELP rfc39_team_sync_errors Total team errors
# TYPE rfc39_team_sync_errors counter
rfc39_team_sync_errors 0
# HELP rfc39_team_sync_invite_pending Total pending team invitations
# TYPE rfc39_team_sync_invite_pending counter
rfc39_team_sync_invite_pending 0
# HELP rfc39_team_sync_noops Total count of noop team sync actions
# TYPE rfc39_team_sync_noops counter
rfc39_team_sync_noops 12
# HELP rfc39_team_sync_previously_invited Total users not invited because we know we invited them already
# TYPE rfc39_team_sync_previously_invited counter
rfc39_team_sync_previously_invited 0
# HELP rfc39_team_sync_removals Total team removals
# TYPE rfc39_team_sync_removals counter
rfc39_team_sync_removals 0
```

grahamc commented:
It looks like the problem is this early return: https://github.com/NixOS/rfc39/blob/c44ecfb2f645242e1e3d6ff97d67eb885210ef33/src/op_sync_team.rs#L198-L202 we probably want to make sure the invitations are saved no matter how we exit this, perhaps breaking the body of the loop out in to another function, and in the existing function call save in all cases? https://github.com/NixOS/rfc39/blob/c44ecfb2f645242e1e3d6ff97d67eb885210ef33/src/op_sync_team.rs#L188-L316 I'm open to many options here, and again I'm happy to help.

grahamc commented:
Locally I applied this diff:
```diff
diff --git a/src/op_sync_team.rs b/src/op_sync_team.rs
index 91a7d06..37a0aac 100644
--- a/src/op_sync_team.rs
+++ b/src/op_sync_team.rs
@@ -228,7 +228,7 @@ pub fn sync_team(
         if let Some(limit) = limit {
             if (additions.get() + removals.get()) >= limit {
                 info!(logger, "Hit maximum change limit");
-                return Ok(());
+                break;
             }
         }
         match action {
```

and am seeing saves correctly. However, I noticed the order is not stable, 

```
[nix-shell:~/nixpkgs]# cat invitations
3105057
335406

[nix-shell:~/nixpkgs]# cat invitations
335406
1699466
3105057

[nix-shell:~/nixpkgs]# cat invitations
3105057
335406
1699466
592849
```

so I further added this patch:

```diff
diff --git a/src/invited.rs b/src/invited.rs
index 4d77e3d..aaa7704 100644
--- a/src/invited.rs
+++ b/src/invited.rs
@@ -78,9 +78,13 @@ impl Invited {
             err
         })?;

-        let string = self
+        let mut values = self
             .invited
             .iter()
+            .collect::<Vec<_>>();
+        values.sort();
+        let string = values
+            .into_iter()
             .map(|id| id.to_string())
             .collect::<Vec<_>>()
             .join("\n");
diff --git a/src/maintainers.rs b/src/maintainers.rs
index 8896cb6..74bd9b4 100644
--- a/src/maintainers.rs
+++ b/src/maintainers.rs
@@ -53,7 +53,7 @@ impl GitHubName {
     }
 }

-#[derive(Debug, PartialEq, Eq, Hash, Copy, Clone, Deserialize)]
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Copy, Clone, Deserialize)]
 pub struct GitHubID(u64);
 impl std::fmt::Display for GitHubID {
```

and it is looking good from here:

```
[nix-shell:~/nixpkgs]# cat invitations
335406
592849
1699466
2379774
3105057
6067895
12898828
```

What do you think about these two patches?

cole-h commented:
LGTM. :shipit:

andresilva committed:
don't early return when maximum change limit is hit

andresilva committed:
sort invited ids when persisting to file

andresilva commented:
@grahamc I applied both your suggestions! 👍 Another alternative for the unstable ordering would have been to just use a `BTreeSet` instead of a `HashSet`.

grahamc commented:
Thanks, merged and applied: https://github.com/NixOS/rfc39-record/commit/d6bca12ebf8af3da0cc27ebb2f7cd1be68f35b57 and it is working. Thank you!

grahamc commented:
Graphs: https://status.nixos.org/grafana/d/w8HHOCBWz/rfc39-maintainer-sync?orgId=1&refresh=30s
